File: problems/Python/10. Regular Expression Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        cache = {}

        def dfs(i, j):
            if (i, j) in cache:
                return cache[(i, j)]
            
            if j == len(p):
                return i == len(s)

            m = i < len(s) and (p[j] == s[i] or p[j] == '.')

            if (j + 1) < len(p) and p[j + 1] == '*':
                res = dfs(i, j + 2) or (m and dfs(i + 1, j))
                cache[(i, j)] = res
                return res

            if m:
                res = dfs(i + 1, j + 1)
                cache[(i, j)] = res
                return res

            cache[(i, j)] = False
            return False

        return dfs(0, 0)

File: problems/Rust/10. Regular Expression Matching.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_match(s: String, p: String) -> bool {
        let mut cache = HashMap::new();
        Self::dfs(0, 0, s.as_bytes(), p.as_bytes(), &mut cache)
    }

    fn dfs(i: usize, j: usize, s: &[u8], p: &[u8], cache: &mut HashMap<(usize, usize), bool>) -> bool {
        if let Some(&res) = cache.get(&(i, j)) {
            return res;
        }

        if j == p.len() {
            return i == s.len();
        }

        let m = i < s.len() && (p[j] == s[i] || p[j] == b'.');
        let res;

        if j + 1 < p.len() && p[j + 1] == b'*' {
            res = Self::dfs(i, j + 2, s, p, cache) || (m && Self::dfs(i + 1, j, s, p, cache));
        } else if m {
            res = Self::dfs(i + 1, j + 1, s, p, cache);
        } else {
            res = false;
        }

        cache.insert((i, j), res);
        res
    }
}